src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp

```diff
@@ -0,0 +1,113 @@
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXGatherElementsOpAdaptor operandAdaptor(operands);
+    ONNXGatherElementsOp gatherElementsOp = cast<ONNXGatherElementsOp>(op);
+    Location loc = op->getLoc();
+
+    ONNXGatherElementsOpShapeHelper shapeHelper(&gatherElementsOp, &rewriter,
+        krnl::getDenseElementAttributeFromKrnlValue,
+        krnl::loadDenseElementArrayValueAtIndex);
+    auto shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    // Insert an allocation and deallocation for the result of this operation.
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    Value output = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput());
+
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value indices = operandAdaptor.indices();
+    int64_t axis = gatherElementsOp.axis();
+    int64_t dataRank = data.getType().cast<ShapedType>().getRank();
+    int64_t indicesRank = indices.getType().cast<ShapedType>().getRank();
+    int64_t outputRank = outputMemRefType.getShape().size();
+    assert(indicesRank == dataRank && "Input tensors must have the same rank");
+    assert(outputRank == dataRank && "Output rank not equal to data rank");
+
+    // Determine whether all indices are positive constants.
+    bool indicesArePositives = indicesArePositiveConstants(indices);
+
+    // Negative value means counting dimensions from the back.
+    axis = axis < 0 ? axis + dataRank : axis;
+
+    MemRefBoundsIndexCapture dataBounds(data), indicesBounds(indices);
+    DimsExpr dataDims, indicesDims;
+    dataBounds.getDimList(dataDims);
+    indicesBounds.getDimList(indicesDims);
+
+    // Gather elements from the 'data' tensor, store them into the output.
+    //   index = indices[i][j]...[n]
+    //   output[i][j]...[n] = data[i][j]..[index]..[n] (index used at axis dim.)
+    //
+    KrnlBuilder createKrnl(rewriter, loc);
+    ValueRange loopDef = createKrnl.defineLoops(indicesRank);
+    DimsExpr lbs(indicesRank, LiteralIndexExpr(0));
+    createKrnl.iterateIE(loopDef, loopDef, lbs, indicesDims,
+        [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
+          // Insert code inside the loop.
+          IndexExprScope innerLoopScope(createKrnl);
+
+          // Access function for indices and output.
+          DimsExpr accessFct;
+          getIndexExprList<DimIndexExpr>(loopInd, accessFct);
+
+          // Compute index = indices[i][j]...[n]
+          Value indexVal = createKrnl.loadIE(indices, accessFct);
+          IndexExpr index = NonAffineIndexExpr(indexVal);
+
+          // When index may be negative, add axis dim to it.
+          if (!indicesArePositives) {
+            LiteralIndexExpr zero(0);
+            SymbolIndexExpr axisDim(dataDims[axis]);
+            index = index.selectOrSelf(index < zero, index + axisDim);
```

Comment:
I agree that this generally works for all cases. But, it seems that `indices` is often a constant tensor. Should we do constant propagation in advance to produce a new constant tensor whose values are not negative, so the loop here would be efficient. It's fine if you want to do this in another PR for both Gather and GatherElements.

Reply:
OK. The code here is conservative. We can consider that in a different PR.